In [1]:
# Install libraries and SQL base

import pandas as pd
import sqlite3

In [2]:
conn = sqlite3.connect("olist.db")

In [3]:
!pip install --upgrade prettytable

In [4]:
%config SqlMagic.autopandas = True

In [5]:
%config SqlMagic.style = '_DEPRECATED_DEFAULT'

In [6]:
%load_ext sql

In [7]:
%sql sqlite:///olist.db

In [8]:
customers = pd.read_csv("olist_customers_dataset.csv")
geolocation = pd.read_csv("olist_geolocation_dataset.csv")
order_items = pd.read_csv("olist_order_items_dataset.csv")
order_payments = pd.read_csv("olist_order_payments_dataset.csv")
order_reviews = pd.read_csv("olist_order_reviews_dataset.csv")
orders = pd.read_csv("olist_orders_dataset.csv")
products = pd.read_csv("olist_products_dataset.csv")
sellers = pd.read_csv("olist_sellers_dataset.csv")
category_translation = pd.read_csv("product_category_name_translation.csv")

In [9]:
customers.to_sql("olist_customers_dataset", conn, if_exists="replace", index=False)
geolocation.to_sql("olist_geolocation_dataset", conn, if_exists="replace", index=False)
order_items.to_sql("olist_order_items_dataset", conn, if_exists="replace", index=False)
order_payments.to_sql("olist_order_payments_dataset", conn, if_exists="replace", index=False)
order_reviews.to_sql("olist_order_reviews_dataset", conn, if_exists="replace", index=False)
orders.to_sql("olist_orders_dataset", conn, if_exists="replace", index=False)
products.to_sql("olist_products_dataset", conn, if_exists="replace", index=False)
sellers.to_sql("olist_sellers_dataset", conn, if_exists="replace", index=False)
category_translation.to_sql("product_category_name_translation", conn, if_exists="replace", index=False)

71

In [10]:
# Check if table imports worked

pd.read_sql("""
SELECT name
FROM sqlite_master
WHERE type = 'table';
""", conn)

,name
0,olist_customers_dataset
1,olist_geolocation_dataset
2,olist_order_items_dataset
3,olist_order_payments_dataset
4,olist_order_reviews_dataset
5,olist_orders_dataset
6,olist_products_dataset
7,olist_sellers_dataset
8,product_category_name_translation


In [11]:
%%sql

-- Read data --
    
SELECT *
FROM olist_orders_dataset
LIMIT 10;

 * sqlite:///olist.db
Done.


,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date
0,e481f51cbdc54678b7cc49136f2d6af7,9ef432eb6251297304e76186b10a928d,delivered,2017-10-02 10:56:33,2017-10-02 11:07:15,2017-10-04 19:55:00,2017-10-10 21:25:13,2017-10-18 00:00:00
1,53cdb2fc8bc7dce0b6741e2150273451,b0830fb4747a6c6d20dea0b8c802d7ef,delivered,2018-07-24 20:41:37,2018-07-26 03:24:27,2018-07-26 14:31:00,2018-08-07 15:27:45,2018-08-13 00:00:00
2,47770eb9100c2d0c44946d9cf07ec65d,41ce2a54c0b03bf3443c3d931a367089,delivered,2018-08-08 08:38:49,2018-08-08 08:55:23,2018-08-08 13:50:00,2018-08-17 18:06:29,2018-09-04 00:00:00
3,949d5b44dbf5de918fe9c16f97b45f8a,f88197465ea7920adcdbec7375364d82,delivered,2017-11-18 19:28:06,2017-11-18 19:45:59,2017-11-22 13:39:59,2017-12-02 00:28:42,2017-12-15 00:00:00
4,ad21c59c0840e6cb83a9ceb5573f8159,8ab97904e6daea8866dbdbc4fb7aad2c,delivered,2018-02-13 21:18:39,2018-02-13 22:20:29,2018-02-14 19:46:34,2018-02-16 18:17:02,2018-02-26 00:00:00
5,a4591c265e18cb1dcee52889e2d8acc3,503740e9ca751ccdda7ba28e9ab8f608,delivered,2017-07-09 21:57:05,2017-07-09 22:10:13,2017-07-11 14:58:04,2017-07-26 10:57:55,2017-08-01 00:00:00
6,136cce7faa42fdb2cefd53fdc79a6098,ed0271e0b7da060a393796590e7b737a,invoiced,2017-04-11 12:22:08,2017-04-13 13:25:17,None,None,2017-05-09 00:00:00
7,6514b8ad8028c9f2cc2374ded245783f,9bdf08b4b3b52b5526ff42d37d47f222,delivered,2017-05-16 13:10:30,2017-05-16 13:22:11,2017-05-22 10:07:46,2017-05-26 12:55:51,2017-06-07 00:00:00
8,76c6e866289321a7c93b82b54852dc33,f54a9f0e6b351c431402b8461ea51999,delivered,2017-01-23 18:29:09,2017-01-25 02:50:47,2017-01-26 14:16:31,2017-02-02 14:08:10,2017-03-06 00:00:00
9,e69bfb5eb88e0ed6a785585b27e16dbf,31ad1d1b63eb9962463f764d4e6e0c9d,delivered,2017-07-29 11:55:02,2017-07-29 12:05:32,2017-08-10 19:45:24,2017-08-16 17:14:30,2017-08-23 00:00:00


In [12]:
%%sql

-- How many total orders are there? --

SELECT COUNT(*) AS total_orders
FROM olist_orders_dataset;

 * sqlite:///olist.db
Done.


,total_orders
0,99441


In [13]:
%%sql

-- How many total customers are there? --

SELECT COUNT(DISTINCT customer_id) AS unique_customers
FROM olist_customers_dataset;

 * sqlite:///olist.db
Done.


,unique_customers
0,99441


In [14]:
%%sql

-- Check the range of the orders --

SELECT 
    MIN(order_purchase_timestamp) AS first_order,
    MAX(order_purchase_timestamp) AS last_order
FROM olist_orders_dataset;

 * sqlite:///olist.db
Done.


,first_order,last_order
0,2016-09-04 21:15:19,2018-10-17 17:30:18


In [15]:
%%sql

-- Check how many orders there are per status --

SELECT 
    order_status,
    COUNT(*) AS order_count
FROM olist_orders_dataset
GROUP BY order_status
ORDER BY order_count DESC;

 * sqlite:///olist.db
Done.


,order_status,order_count
0,delivered,96478
1,shipped,1107
2,canceled,625
3,unavailable,609
4,invoiced,314
5,processing,301
6,created,5
7,approved,2


In [16]:
%%sql

-- Check total revenue --

SELECT 
    ROUND(SUM(payment_value), 2) AS total_revenue
FROM olist_order_payments_dataset;

 * sqlite:///olist.db
Done.


,total_revenue
0,16008872.12


In [17]:
%%sql

-- Check how many transactions there are per payment type, and how much revenue is generated per payment type --

SELECT 
    payment_type,
    COUNT(*) AS transactions,
    ROUND(SUM(payment_value), 2) AS revenue
FROM olist_order_payments_dataset
GROUP BY payment_type
ORDER BY revenue DESC;

 * sqlite:///olist.db
Done.


,payment_type,transactions,revenue
0,credit_card,76795,12542084.19
1,boleto,19784,2869361.27
2,voucher,5775,379436.87
3,debit_card,1529,217989.79
4,not_defined,3,0.00


In [18]:
%%sql

-- Calculate the average order price --

SELECT 
    ROUND(AVG(order_total), 2) AS average_order_value
FROM (
    SELECT 
        order_id,
        SUM(payment_value) AS order_total
    FROM olist_order_payments_dataset
    GROUP BY order_id
);

 * sqlite:///olist.db
Done.


,average_order_value
0,160.99


In [19]:
%%sql

-- Calculate how much revenue is generated per month --

SELECT
    STRFTIME('%Y-%m', o.order_purchase_timestamp) AS month,
    ROUND(SUM(p.payment_value), 2) AS revenue
FROM olist_orders_dataset o
JOIN olist_order_payments_dataset p
    ON o.order_id = p.order_id
GROUP BY month
ORDER BY month;

 * sqlite:///olist.db
Done.


,month,revenue
0,2016-09,252.24
1,2016-10,59090.48
2,2016-12,19.62
3,2017-01,138488.04
4,2017-02,291908.01
5,2017-03,449863.60
6,2017-04,417788.03
7,2017-05,592918.82
8,2017-06,511276.38
9,2017-07,592382.92


In [20]:
%%sql

-- Calculate the amount of orders and revenue per category --

SELECT
    COALESCE(t.product_category_name_english, p.product_category_name)
        AS category,
    COUNT(DISTINCT oi.order_id) AS orders,
    ROUND(SUM(oi.price), 2) AS revenue
FROM olist_order_items_dataset oi
JOIN olist_products_dataset p
    ON oi.product_id = p.product_id
LEFT JOIN product_category_name_translation t
    ON p.product_category_name = t.product_category_name
GROUP BY category
ORDER BY revenue DESC
LIMIT 20;

 * sqlite:///olist.db
Done.


,category,orders,revenue
0,health_beauty,8836,1258681.34
1,watches_gifts,5624,1205005.68
2,bed_bath_table,9417,1036988.68
3,sports_leisure,7720,988048.97
4,computers_accessories,6689,911954.32
5,furniture_decor,6449,729762.49
6,cool_stuff,3632,635290.85
7,housewares,5884,632248.66
8,auto,3897,592720.11
9,garden_tools,3518,485256.46


In [21]:
%%sql

-- Find the top 5 highest value customers --

WITH customer_spend AS (
    SELECT
        c.customer_unique_id,
        SUM(p.payment_value) AS total_spend
    FROM olist_customers_dataset c
    JOIN olist_orders_dataset o
        ON c.customer_id = o.customer_id
    JOIN olist_order_payments_dataset p
        ON o.order_id = p.order_id
    GROUP BY c.customer_unique_id
)

SELECT
    customer_unique_id,
    ROUND(total_spend, 2) AS total_spend,
    CASE
        WHEN total_spend >= 1000 THEN 'High Value'
        WHEN total_spend >= 500 THEN 'Average Value'
        ELSE 'Low Value'
    END AS customer_value
FROM customer_spend
ORDER BY total_spend DESC
LIMIT 5;

 * sqlite:///olist.db
Done.


,customer_unique_id,total_spend,customer_value
0,0a0a92112bd4c708ca5fde585afaa872,13664.08,High Value
1,46450c74a0d8c5ca9395da1daac6c120,9553.02,High Value
2,da122df9eeddfedc1dc1f5349a1a690c,7571.63,High Value
3,763c8b1c9c68a0229c42c9fc6f662b93,7274.88,High Value
4,dc4802a71eae9be1dd28f5d788ceb526,6929.31,High Value


In [22]:
%%sql

-- Calculate the average review score --

SELECT
    ROUND(AVG(review_score), 2) AS average_review_score
FROM olist_order_reviews_dataset;

 * sqlite:///olist.db
Done.


,average_review_score
0,4.09


In [23]:
%%sql

-- Calculate average delivery time per review score --

SELECT
    r.review_score,
    COUNT(*) AS orders,
    ROUND(
        AVG(
            JULIANDAY(o.order_delivered_customer_date)
            - JULIANDAY(o.order_purchase_timestamp)
        ),
        2
    ) AS avg_delivery_days
FROM olist_order_reviews_dataset r
JOIN olist_orders_dataset o
    ON r.order_id = o.order_id
WHERE o.order_delivered_customer_date IS NOT NULL
GROUP BY r.review_score
ORDER BY r.review_score;

 * sqlite:///olist.db
Done.


,review_score,orders,avg_delivery_days
0,1,9409,21.31
1,2,2941,16.66
2,3,7962,14.26
3,4,18987,12.31
4,5,57060,10.69


In [24]:
%%sql

-- Find which categories produce the highest revenue and the lowest revenue--

WITH category_revenue AS (
    SELECT
        COALESCE(t.product_category_name_english, p.product_category_name)
            AS category,
        SUM(oi.price) AS revenue
    FROM olist_order_items_dataset oi
    JOIN olist_products_dataset p
        ON oi.product_id = p.product_id
    LEFT JOIN product_category_name_translation t
        ON p.product_category_name = t.product_category_name
    GROUP BY category
)

SELECT
    category,
    ROUND(revenue, 2) AS revenue,
    RANK() OVER (ORDER BY revenue DESC) AS revenue_rank
FROM category_revenue
ORDER BY revenue_rank;

 * sqlite:///olist.db
Done.


,category,revenue,revenue_rank
0,health_beauty,1258681.34,1
1,watches_gifts,1205005.68,2
2,bed_bath_table,1036988.68,3
3,sports_leisure,988048.97,4
4,computers_accessories,911954.32,5
...,...,...,...
69,flowers,1110.04,70
70,home_comfort_2,760.27,71
71,cds_dvds_musicals,730.00,72
72,fashion_childrens_clothes,569.85,73


In [25]:
%%sql

-- Calculate the lowest, average, and highest delivery times --

SELECT
    ROUND(AVG(
        JULIANDAY(order_delivered_customer_date)
        - JULIANDAY(order_purchase_timestamp)
    ), 2) AS avg_delivery_days,

    ROUND(MIN(
        JULIANDAY(order_delivered_customer_date)
        - JULIANDAY(order_purchase_timestamp)
    ), 2) AS fastest_delivery,

    ROUND(MAX(
        JULIANDAY(order_delivered_customer_date)
        - JULIANDAY(order_purchase_timestamp)
    ), 2) AS slowest_delivery

FROM olist_orders_dataset
WHERE order_delivered_customer_date IS NOT NULL;

 * sqlite:///olist.db
Done.


,avg_delivery_days,fastest_delivery,slowest_delivery
0,12.56,0.53,209.63


## Conclusion

- Analysed 99,441 orders between September 2016 and October 2018, generating R$16.01 million in total revenue with an average order value of 160.99.

- 96,478 orders were delivered, showing a high overall delivery completion rate, while cancellations, unavailable orders, and other incomplete statuses represented a smaller proportion of orders.

- Revenue generally increased throughout 2017 and 2018, with November 2017 generating the highest monthly revenue at approximately R$1.19M.

- September and October 2018 should not be interpreted as genuine revenue declines, as the dataset ends on 17 October 2018, making both months incomplete.

- Health & Beauty, Watches & Gifts, and Bed Bath Table were the three highest-revenue product categories, indicating strong demand in these areas.

- Credit cards were the dominant payment method, generating approximately R$12.54M in revenue and accounting for the majority of transactions.

- Customer analysis identified high-value customers, with the highest-spending customer generating R$13,664.08 in total spend.

- Customer satisfaction was strongly associated with delivery performance: 1-star reviews averaged 21.31 days for delivery, compared with 10.69 days for 5-star reviews.

- Overall delivery time averaged 12.56 days, but the maximum recorded delivery time was 209.63 days, highlighting potential extreme delivery issues.

## Actionable insights

- Improve delivery performance by investigating sellers, regions and logistics processes associated with longer delivery times.

- Prioritise high-value customer retention through loyalty programmes, personalised offers and targeted marketing.

- Invest in high-performing categories such as Health & Beauty, Watches & Gifts, and Bed Bath Table through inventory planning and marketing.

- Investigate low-performing categories before discontinuing them to determine whether low revenue is caused by demand, pricing, availability or limited marketing.

- Monitor cancelled, unavailable and delayed orders to identify operational problems and potential lost revenue.

- Optimise the credit-card checkout experience, given its dominant share of transactions and revenue.

- Use complete months when comparing revenue trends and clearly flag partial periods to avoid misleading conclusions.

- Overall, the analysis suggests that delivery reliability, customer retention and investment in high-performing product categories are the strongest opportunities for improving e-commerce performance.